crates/resctrl/src/lib.rs

```diff
@@ -1130,4 +1184,143 @@ mod tests {
     }
+
+    #[test]
+    fn test_cleanup_all_counts_and_removes() {
+        let fs = MockFs::default();
+        // Simulate mounted
+        fs.add_file(
+            Path::new("/proc/mounts"),
+            "resctrl /sys/fs/resctrl resctrl rw 0 0\n",
+        );
+
+        let root = PathBuf::from("/sys/fs/resctrl");
+        fs.add_dir(Path::new("/sys"));
+        fs.add_dir(Path::new("/sys/fs"));
+        fs.add_dir(&root);
+        // root children
+        fs.add_dir(&root.join("pod_u1"));
+        fs.add_dir(&root.join("pod_u2"));
+        fs.add_dir(&root.join("info"));
+        fs.add_dir(&root.join("mon_groups"));
+        // mon_groups children
+        fs.add_dir(&root.join("mon_groups").join("pod_u1"));
+        fs.add_dir(&root.join("mon_groups").join("pod_u3"));
+        fs.add_dir(&root.join("mon_groups").join("custom"));
+
+        let rc = Resctrl::with_provider(
+            fs.clone(),
+            Config {
+                root: root.clone(),
+                group_prefix: "pod_".into(),
+            },
+        );
+
+        let rep = rc.cleanup_all().expect("cleanup ok");
+        assert_eq!(rep.removed, 4);
+        assert_eq!(rep.removal_failures, 0);
+        assert_eq!(rep.removal_race, 0);
+        // info + mon_groups under root, and custom under mon_groups
+        assert_eq!(rep.non_prefix_groups, 3);
+
+        // Verify removals on fs
+        assert!(!fs.dir_exists(&root.join("pod_u1")));
+        assert!(!fs.dir_exists(&root.join("pod_u2")));
+        assert!(!fs.dir_exists(&root.join("mon_groups").join("pod_u1")));
+        assert!(!fs.dir_exists(&root.join("mon_groups").join("pod_u3")));
+
+        // Non-prefix remain
+        assert!(fs.dir_exists(&root.join("info")));
+        assert!(fs.dir_exists(&root.join("mon_groups")));
+        assert!(fs.dir_exists(&root.join("mon_groups").join("custom")));
+    }
+
+    #[test]
+    fn test_cleanup_all_failures_and_race() {
+        let fs = MockFs::default();
+        // Simulate mounted
+        fs.add_file(
+            Path::new("/proc/mounts"),
+            "resctrl /sys/fs/resctrl resctrl rw 0 0\n",
+        );
+
+        let root = PathBuf::from("/sys/fs/resctrl");
+        fs.add_dir(Path::new("/sys"));
+        fs.add_dir(Path::new("/sys/fs"));
+        fs.add_dir(&root);
+        fs.add_dir(&root.join("mon_groups"));
+
+        // We'll simulate listing that includes one existing, one missing (race),
+        // and one no-permission path for removal.
+        fs.add_dir(&root.join("pod_keep")); // non-prefix
+        fs.add_dir(&root.join("pod_fail"));
+        fs.set_no_perm_remove_dir(&root.join("pod_fail"));
+        // Race: include pod_race in listing override but do not create it
+        fs.set_child_dirs_override(
+            &root,
+            vec![
+                "pod_fail".into(),
+                "pod_race".into(),
+                "x_other".into(), // non-prefix
+            ],
+        );
```

Comment:
let's add another group that will be removed in the root. since the internal implementation calls cleanup_in_dir, let's make sure we pass a mixture of succeeding and failing removals in the same call to cleanup_in_dir (by adding a successful remove in the root)